src/btree/bt_curnext.c

```diff
@@ -38,6 +42,7 @@ __cursor_fix_append_next(WT_CURSOR_BTREE *cbt, bool newpage)
 	 */
 	__cursor_set_recno(cbt, cbt->recno + 1);
 
+retry_insert:
```

Comment:
I think that works (and maybe that's how I originally did it), but thinking about it now, I'd go with:
```
--- ../btree/bt_curnext.c       2019-02-15 17:00:05.955313586 -0500
+++ -   2019-02-15 17:00:17.121389922 -0500
@@ -63,7 +63,7 @@
                cbt->v = 0;
                cbt->iface.value.data = &cbt->v;
        } else {
-               WT_RET(__wt_txn_read(session, cbt->ins->upd, &upd));
+retry_insert:  WT_RET(__wt_txn_read(session, cbt->ins->upd, &upd));
                if (upd == NULL) {
                        cbt->v = 0;
                        cbt->iface.value.data = &cbt->v;
```
That makes more sense to me: `__wt_txn_read()` is the function that caused the problem, the state should be as we left it before that call, so we'll repeat the call rather than re-evaluate anything else about our current state. I think it's the pragmatic choice too -- yes, if we repeat the code we should end up at the same spot, but why chance that?

This comment applies to some of the other edits as well, of course.